### PR TITLE
Word2Vec/Doc2Vec offer model-minimization method Fix issue #446

### PR DESCRIPTION
**Pull request by deepmipt-VladZhukov**
add finished_training method

### CODE REVIEW

File: gensim/models/doc2vec.py [hunk 1]
```diff
@@ -778,6 +778,19 @@ def __str__(self):
             segments.append('t%d' % self.workers)
         return '%s(%s)' % (self.__class__.__name__, ','.join(segments))
 
+    def finished_training(self):
+        """
+        Discard parametrs that are used in training and score. Use if you're sure you're done training a model,
+        """
+        self.training_finished = True
```
Comment (tmylk) [inline gensim/models/doc2vec.py]: Please call the super method in word2vec explicitly

File: gensim/test/test_word2vec.py [hunk 2]
```diff
@@ -482,6 +482,20 @@ def models_equal(self, model, model2):
         most_common_word = max(model.vocab.items(), key=lambda item: item[1].count)[0]
         self.assertTrue(numpy.allclose(model[most_common_word], model2[most_common_word]))
 
+    def testFinishedTraining(self):
+        """Test word2vec model after finishing training"""
+        for i in [0, 1]:
+            for j in [0, 1]:
+                model = word2vec.Word2Vec(sentences, size=10, min_count=0, seed=42, hs=i, negative=j)
+                model.finished_training()
+                self.assertTrue(len(model.vocab), 12)
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: Please tests that necessary attributes are indeed deleted

File: gensim/test/test_doc2vec.py [hunk 3]
```diff
@@ -280,6 +280,14 @@ def models_equal(self, model, model2):
         self.assertEqual(len(model.docvecs.offset2doctag), len(model2.docvecs.offset2doctag))
         self.assertTrue(np.allclose(model.docvecs.doctag_syn0, model2.docvecs.doctag_syn0))
 
+    def test_finished_training(self):
+        """Test doc2vec model after finishing training"""
+        for i in [0, 1]:
+            for j in [0, 1]:
+                model = doc2vec.Doc2Vec(sentences, size=5, min_count=1, negative=i, hs=j)
+                model.finished_training()
+                self.assertTrue(len(model.infer_vector(['graph'])), 5)
```
Comment (tmylk) [inline gensim/test/test_doc2vec.py]: Please tests that necessary attributes are indeed deleted

File: gensim/models/doc2vec.py [hunk 4]
```diff
@@ -392,6 +392,7 @@ def init_sims(self, replace=False):
         etc., but not `train` or `infer_vector`.
 
         """
+        print ('HELLO DOC!!!')
```
Comment (deepmipt-VladZhukov) [inline gensim/models/doc2vec.py]: deleted in next commit

File: gensim/models/word2vec.py [hunk 5]
```diff
@@ -1750,6 +1750,24 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
...
+        if hasattr(self, 'syn1') and not save_syn1:
+            del self.syn1
+        if hasattr(self, 'syn1neg') and not save_syn1neg:
+            del self.syn1neg
+        if hasattr(self, 'syn0_lockf') and not save_syn0_lockf:
+            del self.syn0_lockf
+
+    def finished_training(self):
+        """
+        Discard parametrs that are used in training and score. Use if you're sure you're done training a model.
```
Comment (gojomo) [inline gensim/models/word2vec.py]: typo `parameters`

File: gensim/models/word2vec.py [hunk 6]
```diff
@@ -1750,6 +1750,24 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
...
+        """
+        for i in xrange(self.syn0.shape[0]):
+            self.syn0[i, :] /= sqrt((self.syn0[i, :] ** 2).sum(-1))
+        self.syn0norm = self.syn0
```
Comment (gojomo) [inline gensim/models/word2vec.py]: Not all post-training applications want the unit-normalized vectors!

File: gensim/models/doc2vec.py [hunk 7]
```diff
@@ -778,6 +778,15 @@ def __str__(self):
             segments.append('t%d' % self.workers)
         return '%s(%s)' % (self.__class__.__name__, ','.join(segments))
 
+    def finished_training(self):
+        """
+        Discard parametrs that are used in training and score. Use if you're sure you're done training a model.
+        """
+        self._minimize_model(self.hs, self.negative > 0, True)
+        if hasattr(self, 'doctag_syn0'):
+            del self.doctag_syn0
```
Comment (gojomo) [inline gensim/models/doc2vec.py]: Many will consider the bulk-trained doctag-vectors a part of the model they want to retain.

**Comment by gojomo** [comment]
Your changes closely match the motivating issue, #446 - but even though I originally wrote that, what I know since them makes me think minimization needs to be finer-grained, because much of this state is still relevant for downstream applications even without continued training. So I've added revisions to #446 that echo my line-by-line comments here.

File: gensim/models/word2vec.py [hunk 8]
```diff
@@ -465,7 +465,7 @@ def __init__(
         self.total_train_time = 0
         self.sorted_vocab = sorted_vocab
         self.batch_words = batch_words
-
+        self.training_finished = False
```
Comment (tmylk) [inline gensim/models/word2vec.py]: A better name would be `model_trimmed_post_training = False`

File: gensim/models/word2vec.py [hunk 9]
```diff
@@ -1750,6 +1752,27 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
     def __str__(self):
         return "%s(vocab=%s, size=%s, alpha=%s)" % (self.__class__.__name__, len(self.index2word), self.vector_size, self.alpha)
 
+    def _minimize_model(self, save_syn1 = False, save_syn1neg = False, save_syn0_lockf = False):
+        self.training_finished = True
```
Comment (tmylk) [inline gensim/models/word2vec.py]: Flag is best set in the end of the method

File: gensim/models/word2vec.py [hunk 10]
```diff
@@ -1750,6 +1752,27 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
...
+
+    def discard_model_parameters(self, replace=False):
+        """
+        Discard parameters that are used in training and score. Use if you're sure you're done training a model.
+        If `replace` is set, forget the original vectors and only keep the normalized
+        ones = saves lots of memory!
+        """
+        if replace:
+            for i in xrange(self.syn0.shape[0]):
+                self.syn0[i, :] /= sqrt((self.syn0[i, :] ** 2).sum(-1))
```
Comment (tmylk) [inline gensim/models/word2vec.py]: why duplicate code and not just call `init_sims`?

File: gensim/test/test_doc2vec.py [hunk 11]
```diff
@@ -280,6 +280,24 @@ def models_equal(self, model, model2):
...
+    def test_discard_model_parameters(self):
+        """Test doc2vec model after discard_model_parameters"""
+        for i in [0, 1]:
+            for j in [0, 1]:
+                model = doc2vec.Doc2Vec(sentences, size=5, min_count=1, hs=i, negative=j)
+                model.discard_model_parameters(remove_doctags_vectors=True)
+                self.assertTrue(len(model['human']), 10)
+                self.assertTrue(model.vocab['graph'].count, 5)
+                if (i == 1):
+                    self.assertTrue(hasattr(model, 'syn1'))
```
Comment (tmylk) [inline gensim/test/test_doc2vec.py]: should we assert here that `syn1` is deleted by `_minimize_model`?

File: gensim/models/word2vec.py [hunk 12]
```diff
@@ -1750,6 +1752,25 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
...
+    def _minimize_model(self, save_syn1 = False, save_syn1neg = False, save_syn0_lockf = False):
+        if hasattr(self, 'syn1') and not save_syn1:
+            del self.syn1
+        if hasattr(self, 'syn1neg') and not save_syn1neg:
+            del self.syn1neg
+        if hasattr(self, 'syn0_lockf') and not save_syn0_lockf:
+            del self.syn0_lockf
+        self.model_trimmed_post_training = True
+
+    def discard_model_parameters(self, replace=False):
```
Comment (tmylk) [inline gensim/models/word2vec.py]: `delete_temporary_training_data` maybe is a better name. what do you think?

File: gensim/test/test_doc2vec.py [hunk 13]
```diff
@@ -280,6 +280,24 @@ def models_equal(self, model, model2):
...
+                else:
+                    self.assertTrue(not hasattr(model, 'syn1'))
+                if (j == 1):
+                    self.assertTrue(hasattr(model, 'syn1neg'))
+                else:
+                    self.assertTrue(not hasattr(model, 'syn1neg'))
+                self.assertTrue(hasattr(model, 'syn0_lockf'))
+
```
Comment (deepmipt-VladZhukov) [inline gensim/test/test_doc2vec.py]: Seems I "sync' in git without "commit", when I added self.docvecs, 'doctag_syn0' checks :) will fix it

File: gensim/models/word2vec.py (hunk 12 above)
Comment (deepmipt-VladZhukov) [inline gensim/models/word2vec.py]: My English language skills allows me to only agree with you.

Comment (deepmipt-VladZhukov) [inline gensim/models/word2vec.py]: But in this question obviously yes

File: gensim/test/test_doc2vec.py [hunk 14]
```diff
@@ -280,23 +280,29 @@ def models_equal(self, model, model2):
...
-    def test_discard_model_parameters(self):
-        """Test doc2vec model after discard_model_parameters"""
+    def test_delete_temporary_training_data(self):
+        """Test doc2vec model after delete_temporary_training_data"""
         for i in [0, 1]:
             for j in [0, 1]:
-                model = doc2vec.Doc2Vec(sentences, size=5, min_count=1, hs=i, negative=j)
-                model.discard_model_parameters(remove_doctags_vectors=True)
+                if i == 0 and j == 0:
+                    continue
```
Comment (tmylk) [inline gensim/test/test_doc2vec.py]: we can actually do hs and negative sampling...

File: gensim/test/test_doc2vec.py [hunk 15]
```diff
@@ -280,23 +280,29 @@ def models_equal(self, model, model2):
...
+                model = doc2vec.Doc2Vec(sentences, size=5, min_count=1, window=4, hs=i, negative=j)
```
Comment (tmylk) [inline gensim/test/test_doc2vec.py]: add asserts that it has all the attributes that are about to be deleted

File: gensim/test/test_word2vec.py [hunk 16]
```diff
@@ -482,22 +482,24 @@ def models_equal(self, model, model2):
...
         self.assertTrue(numpy.allclose(model[most_common_word], model2[most_common_word]))
 
-    def testDiscardModelParameters(self):
-        """Test word2vec model after discard_model_parameters"""
+    def testDeleteTemporaryTrainingData(self):
+        """Test word2vec model after delete_temporary_training_data"""
         for i in [0, 1]:
             for j in [0, 1]:
                 model = word2vec.Word2Vec(sentences, size=10, min_count=0, seed=42, hs=i, negative=j)
-                model.discard_model_parameters(replace=True)
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: add assert that it has the attributes that are about to get deleted

**Comment by tmylk** [comment]
Please add a note in CHANGELOG.md describing the change.
It should be for the next release

File: gensim/models/word2vec.py [hunk 17]
```diff
@@ -1751,6 +1753,25 @@ def accuracy(self, questions, restrict_vocab=30000, most_similar=most_similar, c
...
+    def _minimize_model(self, save_syn1 = False, save_syn1neg = False, save_syn0_lockf = False):
+        if hasattr(self, 'syn1') and not save_syn1:
+            del self.syn1
+        if hasattr(self, 'syn1neg') and not save_syn1neg:
+            del self.syn1neg
+        if hasattr(self, 'syn0_lockf') and not save_syn0_lockf:
+            del self.syn0_lockf
+        self.model_trimmed_post_training = True
+
+    def delete_temporary_training_data(self, replace=False):
```
Comment (tmylk) [inline gensim/models/word2vec.py]: Can we rename the `replace` parameter to `replace_word_vectors_with_normalized`?

Comment (deepmipt-VladZhukov) [inline gensim/models/word2vec.py]: I called the parameter this way because we have init_sims(replace=False), with the parameter of the same idea. Should we rename parameter of init_sims to?

**Comment by tmylk** [comment]
Thanks for the PR!
